src/main/java/org/opensearch/knn/index/fetch/KNNFetchSubPhase.java

```diff
@@ -0,0 +1,125 @@
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.index.mapper.ValueFetcher;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.mapper.KNNVectorFieldMapper;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.fetch.FetchContext;
+import org.opensearch.search.fetch.FetchSubPhase;
+import org.opensearch.search.fetch.FetchSubPhaseProcessor;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.knn.common.KNNConstants.BYTES_PER_KILOBYTES;
+
+/**
+ * Fetch sub phase which pull data from doc values.
+ * and fulfill the value into source map
+ */
+@Log4j2
+public class KNNFetchSubPhase implements FetchSubPhase {
+
+    @Override
+    public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) throws IOException {
+        IndexSettings indexSettings = fetchContext.getIndexSettings();
+        if (!KNNSettings.isKNNSyntheticEnabled(indexSettings)) {
+            log.debug("Synthetic is disabled");
+            return null;
+        }
+        MapperService mapperService = fetchContext.mapperService();
+
+        List<DocValueField> fields = new ArrayList<>();
+        for (MappedFieldType mappedFieldType : mapperService.fieldTypes()) {
+            if (mappedFieldType != null && mappedFieldType instanceof KNNVectorFieldMapper.KNNVectorFieldType) {
+                String fieldName = mappedFieldType.name();
+                ValueFetcher fetcher = new DocValueFetcher(
+                    mappedFieldType.docValueFormat(null, null),
+                    fetchContext.searchLookup().doc().getForField(mappedFieldType)
+                );
+                fields.add(new DocValueField(fieldName, fetcher));
+            }
+        }
+        return new KNNFetchSubPhaseProcessor(fetchContext, fields);
+    }
+
+    @AllArgsConstructor
+    @Getter
+    class KNNFetchSubPhaseProcessor implements FetchSubPhaseProcessor {
+
+        private final FetchContext fetchContext;
+        private final List<DocValueField> fields;
+
+        @Override
+        public void setNextReader(LeafReaderContext readerContext) throws IOException {
+            for (DocValueField f : fields) {
+                f.fetcher.setNextReader(readerContext);
+            }
+        }
+
+        @Override
+        public void process(HitContext hitContext) throws IOException {
+            MapperService mapperService = fetchContext.mapperService();
+            final boolean hasNested = mapperService.hasNested();
+            SearchHit hit = hitContext.hit();
+            Map<String, Object> maps = hit.getSourceAsMap();
+            if (maps == null) {
+                // when source is disabled, return
+                return;
+            }
+
+            if (hasNested) {
+                // TODO handle nested field
+                throw new UnsupportedOperationException("knn synthetic source do not support nested field");
```

Comment:
@luyuncheng let make sure that we handle the nested field too. Not handling the nested fields here will push back users to use this capability.

If you want to handle the nested fields case in another PR, we can merge this PR to a feature branch. 

cc: @jmazanec15